### ISSUE TITLE
Unable to get Swarm to connect to VMs on different physical hosts

### ISSUE THREAD

**Issue by codespelunker** [docker/classicswarm#2231]
**TL;DR: When using Swarm with an overlay network, the swarm master is unable to start containers on remote nodes.  I suspect it has something to do with TLS configuration but cannot validate or troubleshoot that problem.  Connecting directly over the network is still successful.**

Docker version: 1.10.3
Swarm version: 1.2.2

On physical **Host A** (_OSX, IP: x.x.x.226_):

```
NAME          ACTIVE   DRIVER       STATE     URL                         SWARM              DOCKER    ERRORS
mh-keystore   -        virtualbox   Running   tcp://192.168.99.113:2376                      v1.11.1
manager       -        virtualbox   Running   tcp://192.168.99.114:2376   manager (master)   v1.11.1
sparse1       *        virtualbox   Running   tcp://192.168.99.115:2376   manager            v1.11.1
```

On physical **Host B** (_Windows, IP: x.x.x.39_):

```
NAME          ACTIVE   DRIVER       STATE     URL                         SWARM              DOCKER    ERRORS
sparse2       -        virtualbox   Running   tcp://192.168.99.108:2376                      v1.11.1 
```

I created my own CA keys (in `/Users/docker/ca`), and from that generated separate keys for manager, sparse1, and sparse2, following the instructions as noted [here](https://docs.docker.com/swarm/configure-tls/).  Manager was created with this:

```
docker-machine \
 --tls-ca-cert "/Users/docker/ca/ca.pem" \
 --tls-ca-key "/Users/docker/ca/ca-priv-key.pem" \
 --tls-client-cert "/Users/docker/ca/manager-cert.pem" \
 --tls-client-key "/Users/docker/ca/manager-priv-key.pem" \
 create -d virtualbox \
 --swarm \
 --swarm-master \
 --swarm-discovery="consul://$(docker-machine ip mh-keystore):8500" \
 --engine-opt="cluster-store=consul://$(docker-machine ip mh-keystore):8500" \
 --engine-opt="cluster-advertise=eth1:2376"
 manager
```

Sparse1 (**Host A**) was created with this:

```
docker-machine \
 --tls-ca-cert "/Users/docker/ca/ca.pem" \
 --tls-ca-key "/Users/docker/ca/ca-priv-key.pem" \
 --tls-client-cert "/Users/docker/ca/sparse1-cert.pem" \
 --tls-client-key "/Users/docker/ca/sparse1-priv-key.pem" \
 create -d virtualbox \
 --swarm \
 --swarm-discovery="consul://$(docker-machine ip mh-keystore):8500" \
 --engine-opt="cluster-store=consul://$(docker-machine ip mh-keystore):8500" \
 --engine-opt="cluster-advertise=eth1:2376" \
 --engine-label="environment=sparse" \
 sparse1
```

Sparse2 (**Host B**) was created with this:

```
docker-machine \
 --tls-ca-cert "c:\Users\docker\ca\ca.pem" \
 --tls-ca-key "c:\Users\docker\ca\ca-priv-key.pem" \
 --tls-client-cert "c:\Users\docker\ca\sparse2-cert.pem" \
 --tls-client-key "c:\Users\docker\ca\sparse2-priv-key.pem" \
 create -d virtualbox \
 --swarm \
 --swarm-discovery="consul://x.x.x.226:8500" \
 --engine-opt="cluster-store=consul://x.x.x.226:8500" \
 --engine-opt="cluster-advertise=x.x.x.39:2376" \
 --engine-label="environment=windows" \
 sparse2
```

My overlay network has been created (called `my-net`) and is visible on all nodes.

On Host A, in the VM manager I have mapped the following:

```
x.x.x.226:8500 -> mh-keystore:8500
x.x.x.226:3376 -> manager:3376
```

On Host B, in the VM manager I have mapped the following:

`x.x.x.39:2376 -> sparse2:2376`

On Host A, my environment is pointed to the swarm master as follows:

```
DOCKER_HOST=192.168.99.114:3376
DOCKER_MACHINE_NAME=manager
DOCKER_TLS_VERIFY=1
DOCKER_CERT_PATH=/Users/rynelson/.docker/machine/machines/manager
```

I have verified that the keys in `/Users/rynelson/.docker/machine/machines/manager` are identical to those in `/Users/docker/ca/manager*`

**Attempt #1: From Host A, run container on local swarm node**
`docker run --rm -e constraint:environment==sparse --net=my-net artifactory/example`

This succeeds, and starts up the example container on sparse1 (Host A).

**Attempt #2: From Host A, run container on remote swarm node**
`docker run --rm -e constraint:environment==windows --net=my-net artifactory/example`

This fails with the following message:

```
docker: Error response from daemon: Unable to find a node that satisfies the following conditions
[available container slots]
[environment==windows].
```

**Attempt #3: From Host A, circumvent swarm and run container on remote node directly**

`docker -H 172.31.66.39:2376 run --rm --net=my-net artifactory/example`

This fails with the message: 

```
docker: An error occurred trying to connect: Post https://x.x.x.39:2376/v1.22/containers/create: x509: certificate is valid for 192.168.99.108, not x.x.x.39.
See 'docker run --help'.
```

I have tried this both with a certificate with no CN, and also a certificate with CN=x.x.x.39.  The result is identical.

**Attempt #4: From Host A, circumvent swarm and run container on remote node directly, disabling TLS**

`docker --tlsverify=false -H x.x.x.39:2376 run --rm --net=my-net artifactory/example`

This succeeds.  The Host B (Windows) box successfully pulls the example image and starts up a container.  However, this defeats the purpose of using Swarm and TLS.

Just for fun, I tried all of these commands on Host B as well.  I got identical results: connecting through Swarm to the remote machine always failed, but connecting directly with TLS disabled succeeded.

I suspect that the TLS issue is causing a problem for Swarm, but I can find no way to fix or even validate that assumption.  Running with debug turned on gives no additional information.

**Comment by codespelunker** [comment]
Any help on this issue?  Have I provided too much information for troubleshooting?  I would like to be able to spin up containers on VMs on two different physical hosts, but this doesn't seem to be possible.

**Comment by BillioJeverson** [comment]
Hello , may I ask is this the same problem as if I have 2 different vm created in my vbox and I wanted to connect them together through swarm ? thank you
